### PR TITLE
added generation code and commands

### LINKED ISSUE #130: US53: List users by generation

### LINKED ISSUE #131: US54: List all users of a generation

### PR DESCRIPTION
**Pull request by aej11a**
closes #130  
closes #131

### CODE REVIEW

File: src/main/java/gedcom/models/Individual.java [hunk 1]
```diff
@@ -429,5 +429,20 @@ public boolean hasAncestor(Individual individual) {
         cousins.removeAll(descendants);
         return cousins;
     }
+    
+    public int getGeneration() {
+    	int generation = 0;
+    	
+    	if(this.getParents().size() > 0) {
+    		for(Individual indi : this.getParents()) {
+    			generation = Math.max(generation, indi.getGeneration() + 1);
```
Comment (robertschaedler3) [inline src/main/java/gedcom/models/Individual.java:438]: This recursive call might cause a stack overflow if there is some form of incest in the family. See #134.

File: src/main/java/gedcom/models/Individual.java [hunk 2]
```diff
@@ -429,5 +429,20 @@ public boolean hasAncestor(Individual individual) {
         cousins.removeAll(descendants);
         return cousins;
     }
+    
+    public int getGeneration() {
```
Comment (robertschaedler3) [inline src/main/java/gedcom/models/Individual.java:433]: Is generation calculated from the oldest ancestor of an individual or from the oldest individual in the GEDFile?

File: src/main/java/gedcom/models/Individual.java (hunk 1 above)
Comment (aej11a) [reply src/main/java/gedcom/models/Individual.java:438]: @robertschaedler3 Any ideas how to safeguard against this? I'm having trouble thinking of something effective. Maybe we could just validate No-Incest first, and if the fails then print some error like "can't compute generation with circular relationship" ? Let me know what you think

File: src/main/java/gedcom/models/Individual.java (hunk 2 above)
Comment (aej11a) [reply src/main/java/gedcom/models/Individual.java:433]: @robertschaedler3 I defined it as the oldest ancestor of an individual. I thought about doing it the other way, but I couldn't think of any good way to distinguish generations between (for example) the two oldest couples in a GED File, assuming they're not related (related by marriage of their grandkids for example).

So this is "what generation is Sally within her own family?"

File: src/main/java/gedcom/models/Individual.java (hunk 1 above)
Comment (robertschaedler3) [reply src/main/java/gedcom/models/Individual.java:438]: I solved this issue in #134 for the existing recursive traversals by converting them to iterative solutions that utilize a HashSet to keep track of visited individuals.

File: src/main/java/gedcom/models/Individual.java (hunk 2 above)
Comment (robertschaedler3) [reply src/main/java/gedcom/models/Individual.java:433]: This makes sense to me. Thanks!

**Review by Liam-Brew — APPROVED** [review]

File: src/main/java/gedcom/models/Individual.java (hunk 1 above)
Comment (aej11a) [reply src/main/java/gedcom/models/Individual.java:438]: @robertschaedler3 Do we need any more changes here now that #134 has been merged?

**Review by robertschaedler3 — APPROVED** [review]

File: src/main/java/gedcom/models/Individual.java (hunk 1 above)
Comment (robertschaedler3) [reply src/main/java/gedcom/models/Individual.java:425]: Looks good to me. Resolving this thread.

**Pull request merged**